src/libraries/System.Net.Http.Json/tests/FunctionalTests/HttpClientJsonExtensionsTests.cs

```diff
@@ -441,5 +499,82 @@ public async Task GetFromJsonAsync_EnforcesTimeout(bool useDeleteAsync, bool slo
                 }));
             });
         }
+
+        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsNotBrowser))] // No Socket support
+        [InlineData(true)]
+        [InlineData( false)]
+        public async Task GetFromJsonAsAsyncEnumerable_EnforcesTimeout(bool slowHeaders)
+        {
+            TaskCompletionSource<byte> exceptionThrown = new(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            await LoopbackServer.CreateClientAndServerAsync(async uri =>
+            {
+                using var client = new HttpClient { Timeout = TimeSpan.FromMilliseconds(100) };
+
+                Exception ex = await Assert.ThrowsAsync<TaskCanceledException>(async () =>
+                {
+                    await foreach (string? str in client.GetFromJsonAsAsyncEnumerable<string>(uri))
+                    {
+                        _ = str;
+                    }
+                });
+
+#if NETCOREAPP
+                Assert.Contains("HttpClient.Timeout", ex.Message);
+                Assert.IsType<TimeoutException>(ex.InnerException);
+#endif
+
+                exceptionThrown.SetResult(0);
+            },
+            async server =>
+            {
+                // The client may timeout before even connecting the server
+                await Task.WhenAny(exceptionThrown.Task, Task.Run(async () =>
+                {
+                    try
+                    {
+                        await server.AcceptConnectionAsync(async connection =>
+                        {
+                            if (!slowHeaders)
+                            {
+                                await connection.SendPartialResponseHeadersAsync(headers: new[] { new HttpHeaderData("Content-Length", "42") });
+                            }
+
+                            await exceptionThrown.Task;
+                        });
+                    }
+                    catch { }
+                }));
+            });
+        }
+
+        [Fact]
+        public async Task GetFromJsonAsAsyncEnumerable_CorrectlyExposesDeferredExecutionSemantics()
+        {
+            await LoopbackServer.CreateClientAndServerAsync(async uri =>
+            {
+                using var client = new HttpClient { Timeout = TimeSpan.FromMilliseconds(100) };
+
+                // Get the async enumerable, but don't iterate it - this is used to validate deferred execution.
+                var getAsyncEnumerable = client.GetFromJsonAsAsyncEnumerable<string>(uri);
+
+                // Wait longer than the timeout.
+                await Task.Delay(TimeSpan.FromMilliseconds(150));
```

Comment:
Tests that depend on timing like these tend to become very unreliable in CI.

This might work if you skip the `LoopbackServer` and only emulate the server response.
Something like
```c#
using var client = new HttpClient(new CustomResponseHandler((r, c) =>
{
    return new HttpResponseMessage
    {
        Content = new StringContent("your json response here")
    };
}))
{
    Timeout = TimeSpan.FromMilliseconds(1)
};

await foreach (string s in client.GetFromJsonAsAsyncEnumerable<string>("http://dummyUrl"))
{
    // Wait longer than the timeout.
    await Task.Delay(TimeSpan.FromMilliseconds(10));
}
```